include/alp/reference/blas1.hpp

```diff
@@ -751,22 +767,42 @@ namespace alp {
 		typename IOType, typename IOStructure, typename IOView, typename IOImfR, typename IOImfC,
 		class Monoid
 	>
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const Monoid & monoid = Monoid(),
-		const typename std::enable_if< ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value, void >::type * const = NULL ) {
+		const typename std::enable_if_t<
+			!alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value
+		> * const = NULL
+	) {
 		// static sanity checks
 		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D1, IOType >::value ), "alp::foldl",
-			"called with a vector x of a type that does not match the first domain "
+			"called with a vector y of a type that does not match the first domain "
 			"of the given operator" );
 		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D2, InputType >::value ), "alp::foldl",
 			"called on a vector y of a type that does not match the second domain "
 			"of the given operator" );
```

Comment:
I am not 100% sure about this, but should IOType match D2 and InputType match D1? This operation is doing IO = INPUT . IO

Reply:
yes since in op, D3 = D1.D2 and in foldr D2=D3 (in foldl D1=D3)